### ISSUE TITLE
Slashes are stripped from template part block attributes with WP 6.5 beta

### ISSUE THREAD

**Issue by ktmn** [woothemes/woocommerce#45176]
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

I edited a template part, everything broke. Most of my blocks lost all their attributes for no apparent reason. I have narrowed it down to WP beta and WC. I have managed to reproduce it in WP playground, please follow the steps. 

### Expected behavior

Don't mangle.

### Actual behavior

Mangles.

### Steps to reproduce

1. Go to https://playground.wordpress.net/?wp=beta&plugin=woocommerce (At the time of posting this should load WP 6.5-beta3, WC 8.6.1)

2. From the admin bar, open **My WordPress Website -> Dashboard** in a new tab to break out of the WP playground iframe, skip the WC setup wizard, choose any country.

3. In the browser address bar, append `/wp-admin/site-editor.php?postType=wp_template_part&postId=twentytwentyfour%2F%2Fpost-meta&categoryId=uncategorized&categoryType=wp_template_part&canvas=edit` to the scoped WP playground URL (or make your way to **Appearance -> Editor -> Patterns -> General -> Post Meta** and edit that template part).

4. Delete all content, replace it with the following (`Ctrl+Shift+V`):
```
<!-- wp:paragraph {"metadata":{"name":"c\u002d\u002dd"},"className":"a\u002d\u002db"} -->
<p class="a--b">test</p>
<!-- /wp:paragraph -->
```
This is a paragraph that has a class name `a--b` and block name `c--d`, the `--` string gets stored in block attributes as `\u002d\u002d`.

5. Save the template part.

6. Change the paragraph text from `test` to `test2`, save the template part again.

7. In the right sidebar **Template Part -> Revisions (3)** should now show up, visit it (`/wp-admin/revision.php?revision=18`).

8. When the template part was saved for the second time, it created revisions 2 and 3. 2 changed the text `test` to `test2`, 3 mangled the block attributes `<!-- wp:paragraph {"metadata":{"name":"c\u002d\u002dd"},"className":"a\u002d\u002db"} -->` to `<!-- wp:paragraph {"metadata":{"name":"cu002du002dd"},"className":"au002du002db"} -->`, removing the backslashes. When doing the same procedure without WooCommerce plugin enabled, there will only be 2 revisions, the 3rd one is not created.

When you go back to edit the template part, the block name now displays as `cu002du002dd` when it was supposed to be `c--d`, but furthermore, whenever you edit the template part and save, it creates **2 revisions**. When WC is not enabled, it only creates 1 revision. So what is WC doing?

### WordPress Environment

https://playground.wordpress.net/?wp=beta&plugin=woocommerce (WP 6.5-beta3, WC 8.6.1)

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [X] This bug happens with a default WordPress theme active, or [Storefront](https://woo.com/products/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

**Comment by kmanijak** [comment]
Hi @ktmn,

Thanks for taking the time to raise this issue and especially for narrowing it down to WooCommerce + WordPress 6.5!

I could reproduce it:
- only if I edited the attributes in Code Editor directly,
- only with two consequent dashes and not with single one.
I could not reproduce it when I input the attributes via inspector controls as in a screenshot below.

<img width="290" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/1bc861b2-6ba0-409c-ad3e-c8839ce93418">

Is that your experience as well?

Also, I'd like to better understand the consequences you described:

> I edited a template part, everything broke. Most of my blocks lost all their attributes

What do you mean by "lost all their attributes"? Were some attributes removed?

**Comment by ktmn** [comment]
> only with two consequent dashes and not with single one.

Yes, only 2 consecutive dashes are converted to `\u002d\u002d`, not sure why, but this issue is not about dashes specifically, it's about anything with a backslash in block attributes. For example if you make the Block name `a & b` it will be stored in block attributes as `a \u0026 b` but the bug will mangle the value to `a u0026 b`.

> What do you mean by "lost all their attributes"? Were some attributes removed?

By my understanding the bug removed backslashes from lots of block attributes that had them and Gutenberg found them invalid or something and removed attributes from some blocks altogether. So I found myself with bunch of broken blocks with no attributes at all. But basically I'm just giving context for how I stumbled upon this issue, it's not actually important.

> I could not reproduce it when I input the attributes via inspector controls as in a screenshot below.

I _can_ reproduce it when using only inspector controls. In step 4 I'm giving the block code for pasting only for convenience, you can simply delete all content, add a paragraph, add the `a--b` class, `d--e` block name for good measure, save, change text, save again and then look at the revisions:

![image](https://github.com/woocommerce/woocommerce/assets/6892634/8c929ea1-3704-4ddf-9d3a-32e8ec9cb74b)

There is a 3rd revision that removes the backslashes.

Everything can appear ok in the editor, that's because after saving, the content is not replaced with what it is in the database, but if you were to reload the page/make your way back to edit the template part, it would have "mangled" content.

Something to latch onto here, is the fact that when this bug occurs, whenever you save a template part **2** revisions are created. What is creating the second revision? I scoured the WC source code for `wp_update_post` or similar that could be doing it but couldn't find anything. _Also_ it's only happening with WP beta, so maybe looking at the diff for what has changed in WP side could help.

Are you able to observe the revisions?

![image](https://github.com/woocommerce/woocommerce/assets/6892634/9ae3571f-6de6-4a30-b45e-22adb1d0b4ca)

\*save\*

![image](https://github.com/woocommerce/woocommerce/assets/6892634/3d844691-4db7-4dd8-b1b0-808341465152)

The 2 revisions happen only when there is something to mangle. When I create a simple paragraph with just text and save - 1 revision is created. When I make the paragraph's block name `&` and save - 2 revisions are created:

1.
![image](https://github.com/woocommerce/woocommerce/assets/6892634/8d592f3c-b27b-46b9-accb-8e19e43b2f71)

2.
![image](https://github.com/woocommerce/woocommerce/assets/6892634/1c102779-2948-4ef0-be78-b36ee8e6efef)

---

There is also a long-standing known issue with slashes being removed in post meta: https://wordpress.stackexchange.com/questions/53336/wordpress-is-stripping-escape-backslashes-from-json-strings-in-post-meta
But here a similar thing seems to be happening with `post_content` itself.

**Comment by kmanijak** [comment]
Thanks a lot for providing more details!

We did an investigation and the bug occurs when using Block Hooks API in general. I was able to reproduce it without WooCommerce installed hence I raised a ticket in WordPress: https://core.trac.wordpress.org/ticket/60671 (more details in the ticket). WooCommerce is a consumer of Block Hooks API so you could reproduce the issue with WooCommerce installed.

Having said that I'll go ahead and close this issue and let's follow the [Core Trac ticket](https://core.trac.wordpress.org/ticket/60671) for further updates.

Please note the closed issues can still be commented on.

**Comment by ktmn** [comment]
Can confirm it's fixed in 6.5-RC2 👌
Thank you!
